gdal/alg/viewshed.cpp

```diff
@@ -0,0 +1,612 @@
+ * pixel value which should be ignored in generating contours as if the value
+ * of the pixel were not known.
+ *
+ * @param hLayer The layer to which new contour vectors will be written.
+ * Each contour will have a LINESTRING geometry attached to it.   This
+ * is really of type OGRLayerH, but void * is used to avoid pulling the
+ * ogr_api.h file in here.
+ *
+ * @param iIDField If not -1 this will be used as a field index to indicate
+ * where a unique id should be written for each feature (contour) written.
+ *
+ * @param iElevField If not -1 this will be used as a field index to indicate
+ * where the elevation value of the contour should be written.
+ *
+ * @param pfnProgress A GDALProgressFunc that may be used to report progress
+ * to the user, or to interrupt the algorithm.  May be NULL if not required.
+ *
+ * @param pProgressArg The callback data for the pfnProgress function.
+ *
+ * @return CE_None on success or CE_Failure if an error occurs.
+ */
+
+#define EARTH_DIAMETER 12741994.0;
+
+#ifndef round
+inline double round( double d )
+{
+    return floor( d + 0.5 );
+}
+#endif
+
+
+void SetVisibility(int iPixel, double dfZ, double dfZTarget, double* padfZVal,
+    GByte* pabyResult, GByte byVisibleVal, GByte byInvisibleVal)
+{
+    if (padfZVal[iPixel] + dfZTarget < dfZ)
+    {
+        padfZVal[iPixel] = dfZ;
+        pabyResult[iPixel] = byInvisibleVal;
+    }
+    else
+        pabyResult[iPixel] = byVisibleVal;
+}
+
+int AdjustHeightInRange(double* adfGeoTransform, int iPixel, int iLine, double& dfHeight, double dfDistance2, double dfCurvCoeff)
+{  
+    if (dfDistance2 <= 0 && dfCurvCoeff == 0)
+        return TRUE;
+
+    double dfX = adfGeoTransform[1] * iPixel + adfGeoTransform[2] * iLine;
+    double dfY = adfGeoTransform[4] * iPixel + adfGeoTransform[5] * iLine;
+    double dfR2 = dfX * dfX + dfY * dfY;
+
+    /* calc adjustment */
+    if (dfCurvCoeff != 0)
+        dfHeight -= dfCurvCoeff * dfR2 / EARTH_DIAMETER;
+    
+    if (dfDistance2 > 0 && dfR2 > dfDistance2)
+        return FALSE;
+
+    return TRUE;
+}
+
+double CalcHeightLine(int i, double Za, double Zo)
+{
+    if (i == 1)
+        return Za;
+    else
+        return (Za - Zo) / (i - 1) + Za;
+}
+
+double CalcHeightDiagonal(int i, int j, double Za, double Zb, double Zo)
+{
+    return ((Za - Zo) * i + (Zb - Zo) * j) / (i + j - 1) + Zo;
+}
+
+double CalcHeightEdge(int i, int j, double Za, double Zb, double Zo)
+{
+    if (i == j)
+        return CalcHeightLine(i, Za, Zo);
+    else
+        return ((Za - Zo) * i + (Zb - Zo) * (j - i)) / (j - 1) + Zo;
+}
+
+double CalcHeight(double dfZ, double dfZ2, GDALViewshedMode eMode)
+{
+    if (eMode == GVM_Edge)
+        return dfZ2;
+    else if (eMode == GVM_Max)
+        return std::max(dfZ, dfZ2);
+    else if (eMode == GVM_Min)
+        return std::min(dfZ, dfZ2);
+    else
+        return dfZ;
+}
+
+CPLErr GDALViewshedGenerate(GDALRasterBandH hBand, const char* pszTargetRasterName,
```

Comment:
in addition to pszTargetRasterName, I'd argue for a driver name and creation options so that people can do a MEM dataset, or compress the TIFF